include/envoy/http/filter.h

```diff
@@ -213,6 +213,15 @@ class StreamFilterCallbacks {
    */
   virtual Router::RouteConstSharedPtr route(const Router::RouteCallback& cb) PURE;
 
+  /**
+   * Sets the cached route for the current request to the passed-in RouteConstSharedPtr parameter.
+   *
+   * NOTE: Similar to route(const Router::RouteCallback& cb), this route that is set will be
+   * overridden by clearRouteCache() in subsequent filters. Usage is intended for filters at the end
+   * of the filter chain.
+   */
+  virtual void setRoute(const Router::RouteConstSharedPtr route) PURE;
```

Comment:
What happens if we pass nullptr here? Does that make it equivalent to clearRouteCache? Should we consider disallowing nullptr here?

Reply:
It’s the same as when `snapped_route_config_->route(cb, *request_headers_, filter_manager_.streamInfo(), stream_id_);` returns nullptr when no route match is found, since I essentially grabbed the core logic from the [current refreshCachedRoute implementation](https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_manager_impl.cc#L1252-L1279). So the nullptr case is already accounted for.  

 `clearRouteCache` is slightly different in that it sets the cached attributes to optional ptrs 
  ``` 
  cached_route_ = absl::optional<Router::RouteConstSharedPtr>();
cached_cluster_info_ = absl::optional<Upstream::ClusterInfoConstSharedPtr>();
``` 

Would my `setRoute` API make more sense if I replicate the `clearRouteCache` behavior (setting to optional pointers) when route is nullptr?

Reply:
Apologies, for some reason, Github doesn't resolve my cpp Markdown code block correctly. But hopefully you're able to parse my question :D

Reply:
resolving, will differentiate setRoute(nullptr) from clearRouteCache() per this discussion https://github.com/envoyproxy/envoy/pull/15266#issuecomment-808517011